Vol gradient

corredD opened the pull request:
<!-- Thank you for contributing to Mol* -->

# Description
New representation for volume, show gradient field line as lines or cylinders.

<img width="3354" height="1846" alt="image" src="https://github.com/user-attachments/assets/81eee459-b04a-4d0d-a423-d64ab597a263" />

<img width="487" height="506" alt="image" src="https://github.com/user-attachments/assets/fe79b46d-85fb-493c-a975-1fd74addc5bc" />

## Actions

- [ ] Added description of changes to the `[Unreleased]` section of `CHANGELOG.md`
- [ ] Updated headers of modified files
- [ ] Added my name to `package.json`'s `contributors`
- [ ] (Optional but encouraged) Improved documentation in `docs`

corredD commented on src/mol-repr/volume/gradient.ts:
```diff
@@ -0,0 +1,436 @@
+    maxSteps: PD.Numeric(1000, { min: 1, max: 2000, step: 1 }, { description: 'Maximum number of steps for gradient lines.' }),
+    stepSize: PD.Numeric(0.35, { min: 0.01, max: 10, step: 0.01 }, { description: 'Step size for gradient lines.' }),
+    minSpeed: PD.Numeric(0.001, { min: 0, max: 1, step: 1e-6 }, { description: 'Minimum speed for gradient lines.' }),
+};
```
![disk305x302](https://github.com/user-attachments/assets/0a0463a2-6e6b-4e2a-9053-2d53d178e06e)

arose replied:
I think we could do this fully in the shader by a) having a global time and b) a per vertex attribute value that defines the position on a stream line. With that we can make parts of the stream line transparent based on the global time. We also need a way in the renderer to continuously render not just on changes.

Let's do this after the publications.

corredD commented:
> > @giagitom do you have some time to help get this in?
> 
> @dsehnal I'll try my best but I know nothing about the algorithm itself.. I'll try to look into this during next days
> 
> Anyway @corredD this seems a super nice addition!

Thanks @giagitom ! maybe you can help with the color theme ? it seems that the external volume work better than the volume-value theme. I tried to fix it, not sure I did it right. Did you got proper coloring with the Dot representation ?

arose commented:
Thanks @corredD this looks nice, will review later.

Regarding "pymol algo". Did you copy any code? Because PyMOL is GPL last I checked and we do not want GPL code here.

The isovalue param is not used, right?

corredD commented:
@arose thanks for looking at it! The open-source parts of PyMOL are under a [BSD-style license](https://github.com/schrodinger/pymol-open-source/blob/a550aa709f45f5a57146f3bd57a8b6dcdc4db21c/LICENSE). No copy/paste has been done here — this is an independent reimplementation, with algorithmic inspiration only. But for sake of transparancy I added a comment 'Algorithm based on PyMOL’s isosurface gradient tracing (Schrödinger LLC)'. 
Code was written from the Dot representation with the help of GPT-5, which generates MIT-licensed code by default ( hope its ok with you ). I removed the isoValue param which is not used. I also changed the color-theme, but I am not sure I did it correctly...

dsehnal commented on src/mol-repr/volume/gradient.ts:
```diff
@@ -0,0 +1,726 @@
+
+function traceStreamlineBothDirs(space: Tensor.Space, data: ArrayLike<number>, seed: Vec3, maxSteps: number, dsWorld: number, eps: number, hx: number, hy: number, hz: number, writeStride: number, order: 2 | 4): StreamlinePoint[] {
+    const line: StreamlinePoint[] = [];
+    const S = { p: Vec3(), g: Vec3(), g2: Vec3(), g3: Vec3(), g4: Vec3(), v1: Vec3(), v2: Vec3(), v3: Vec3(), v4: Vec3(), k1: Vec3(), k2: Vec3(), k3: Vec3(), k4: Vec3(), dp: Vec3(), t1: Vec3(), t2: Vec3(), t3: Vec3() };
```
This should be declared outside the function.

dsehnal commented on src/mol-repr/volume/gradient.ts:
```diff
@@ -0,0 +1,726 @@
+    for (let a = 0; a < rangeSize; a++) {
+        const activeCells: number[] = [];
+
+        const walk = (sign: 1 | -1) => {
```
A closure inside a loop is pretty bad for perf.

dsehnal commented on src/mol-repr/volume/gradient.ts:
```diff
@@ -0,0 +1,726 @@
+                const level = interpolateValueFrac(space, data, li, lj, lk, fx, fy, fz);
+                if (level < props.minLevel || level > props.maxLevel) break;
+
+                const gW = Vec3();
```
No vector allocations inside a loop.

dsehnal commented on src/mol-repr/volume/gradient.ts:
```diff
@@ -0,0 +1,726 @@
+
+function clampIndex(space: Tensor.Space, x: number, y: number, z: number) {
+    const nx = space.dimensions[0] as number, ny = space.dimensions[1] as number, nz = space.dimensions[2] as number;
+    return [Math.max(0, Math.min(nx - 1, x)), Math.max(0, Math.min(ny - 1, y)), Math.max(0, Math.min(nz - 1, z))] as [number, number, number];
```
`clampIndex` seems to be called inside a loop -- this should get an `out: [number, number, number]` vector instead as a param too.

corredD commented on src/mol-repr/volume/gradient.ts:
```diff
@@ -0,0 +1,726 @@
+    }, materialId);
+}
+
+export function VolumeCylindersImpostorVisual(materialId: number): VolumeVisual<VolumeCylindersLinesParams> {
```
I added the mesh fallback for spheres and cylinders, I am actually fine with the current cylinder looks. But the tube mesh would definitely look better. I will look into it.

arose commented on src/mol-repr/volume/gradient.ts:
```diff
@@ -0,0 +1,726 @@
+    return Volume.Loci(volume, instances);
+}
+
+function getGradientLoci(pickingId: PickingId, volume: Volume, _key: number, props: VolumeGradientProps, id: number) {
```
Something like `VolumeIsosurfaceTexture` in src\mol-repr\volume\isosurface.ts. It stores things in volume._propertyData. But the streamlines need to be a bit different as they depend on parameters and you can have multiple set of streamlines per volume at the same time.

corredD replied:
I put some caching system in the representation file, but how would you do the Volume.Streamline.Loci ? And whats the issue with the current default one ?

arose replied:
You currently use a streamline index with a Cell Loci which makes no sense. You would need to implement a Volume.Streamline.Loci, similar to Volume.Cell.Loci but using streamline index not cell index. There are also a bunch of place where support for a new Loci needs to be added, best to use an existing Loci as an example.

arose commented on src/mol-theme/color/volume-value.ts:
```diff
@@ -89,11 +93,18 @@ export function VolumeValueColorTheme(ctx: ThemeDataContext, props: PD.Values<Vo
                 if (isNaN(value)) return props.defaultColor;
 
-                return (clamp((value - domain[0]) / (domain[1] - domain[0]), 0, 1) * ColorTheme.PaletteScale) as Color;
+                if (usePalette) {
```
Fixed it in 0e968ae59c08b967413ef755bcc301f29758e53d. You can undo your changes.

giagitom commented:
> I tried to fix it, not sure I did it right. Did you got proper coloring with the Dot representation ?

@corredD in my implementation I am using these parameters (without specifying the offset for each color) and it is working accordingly with volume-value size theme

 ```ts 
       { 
          name : 'volume-value',
          params : {
            colorList : {
              colors : this.palette.colors.map(c => parseInt(c, 16)),
              kind : this.palette.kind
            }
          }
        }
```
